Add file status extension to the LSP module

DavidGoldman opened the pull request:
- For more information see https://forums.swift.org/t/sourcekit-lsp-file-status-ux/35947

DavidGoldman commented:
@swift-ci please test

DavidGoldman commented:
@swift-ci please test

DavidGoldman commented:
@swift-ci please test

benlangmuir commented on Sources/LanguageServerProtocol/Notifications/SourceKitFileStatusNotification.swift:
```diff
@@ -0,0 +1,48 @@
+//
+//===----------------------------------------------------------------------===//
+
+/// Notification from the server about a per-file status update for the client.
```
I've been putting `**LSP Extension**` into the first line of the description so it shows up prominently in quick help/hover.

DavidGoldman replied:
Done

benlangmuir commented on Sources/LanguageServerProtocol/Notifications/SourceKitFileStatusNotification.swift:
```diff
@@ -0,0 +1,48 @@
+  /// Status category, used to stylize the status in the editor (e.g. with an icon).
+  public var category: SourceKitFileStatusCategory
+
+  /// User visible description.
```
Maybe "human readable" instead of "user visible" since whether it's shown in the UI is up to the client?

DavidGoldman replied:
Done

benlangmuir commented on Sources/LanguageServerProtocol/Notifications/SourceKitFileStatusNotification.swift:
```diff
@@ -0,0 +1,48 @@
+
+  /// URI to open on click, if the editor supports it. Can be used to point
+  /// users towards an expanded description or error link.
+  public var clickTarget: DocumentURI?
```
We should avoid terminology like "click".  The client defines the UI, not us, and we should provide information that is useful to a variety of clients.  How about "infoLink" or something similar?

DavidGoldman replied:
Done

benlangmuir commented on Sources/LanguageServerProtocol/SupportTypes/SourceKitFileStatus.swift:
```diff
@@ -0,0 +1,31 @@
+//===----------------------------------------------------------------------===//
+
+/// High level categorization of a file status.
+public enum SourceKitFileStatusCategory: Int, Codable, Hashable {
```
Unless you have a particular reason to choose Int, I would suggest String and to make this a struct with static fields so that we can extend it with new kinds over time.

DavidGoldman replied:
Done

DavidGoldman commented:
> A few suggestions, but overall seems good to me. I think we should merge this at the same time as the functionality itself though. Also, how about adding documentation for this (e.g. an `Extensions.md` file in Documentation).

SGTM, will do both when I get a chance.

DavidGoldman commented:
PTAL, I've updated the design and added a `sourcekit/fileStatus` initialization option to disable/enable it. If all looks well I'll go ahead and add tests + documentation.

In follow up PRs I'll add the following:
- sourcekitd and clangd integration (for AST status)
- VSCode extension support

DavidGoldman commented:
@swift-ci please test

DavidGoldman commented:
@swift-ci please test

benlangmuir commented on Sources/SourceKitLSP/ToolchainLanguageServer.swift:
```diff
@@ -67,3 +71,13 @@ public protocol ToolchainLanguageServer: AnyObject {
 }
+
+/// Options for a `ToolchainLanguageServer`.
+public struct InternalServerOptions {
```
Why not call it `ToolchainLanguageServerOptions`?

benlangmuir commented on Sources/SourceKitLSP/SourceKitServer.swift:
```diff
@@ -427,6 +448,9 @@ extension SourceKitServer {
           log("expected null or int for 'maxResults'; got \(invalid)", level: .warning)
         }
       }
+      if case .bool(let fileStatus) = options["sourcekit/fileStatus"] {
```
Why an initialization option rather than a client capability? This feels more like a capability to me.

akyrtzi commented on Sources/LanguageServerProtocol/Notifications/SourceKitFileStatusNotification.swift:
```diff
@@ -0,0 +1,48 @@
+  public var state: SourceKitFileState
+
+  /// Severity of the status. Used to stylize the status (e.g. with a color or icon).
+  public var severity: DiagnosticSeverity
```
Should this be optional and set to `nil` for the "all-OK state"?

I like the general idea of just classifying as "warning" when it is known that there may be issues. For "error" I'd probably only use it for really catastrophic events like the build service crashing or something.

DavidGoldman replied:
Yeah, I was thinking this would only be used for warnings and errors, I could make it a separate type. Example cases I was thinking:

* Warning: file using fallback build settings or needs prebuilding
* Error: Internal error or invalid build set up (e.g. error in Package.swift, error in BUILD file)

I separated this from the working vs. ready status to denote loading when in the error / warning state (e.g. if reloading Package.swift files). In VS Code, you can use a separate color to handle the severity (or an icon), which makes the warning/error states more pronounced. Let me know which you'd prefer.

DavidGoldman commented on Sources/LanguageServerProtocol/Notifications/SourceKitFileStatusNotification.swift:
```diff
@@ -0,0 +1,48 @@
+  public var severity: DiagnosticSeverity
+
+  /// Human readable description.
+  public var message: String
```
I was just thinking the string would be empty then, but I could make it optional

DavidGoldman replied:
Also: clangd provides a description about the work it is doing; I was also thinking it could be useful to denote build system state such as the following:

* Target the build settings are being resolved from
* SDK platform/version, toolchain version

shahmishal commented:
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create the pull request with `main` branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412
